packages/order-utils/src/rich_reverts.ts

```diff
@@ -0,0 +1,376 @@
+            sender,
+        });
+    }
+}
+
+export class InvalidTakerError extends RichRevertReason {
+    constructor(orderHash?: string, taker?: string) {
+        super(declarationToAbi('InvalidTakerError(bytes32 orderHash, address taker)'), {
+            orderHash,
+            taker,
+        });
+    }
+}
+
+export class InvalidMakerError extends RichRevertReason {
+    constructor(orderHash?: string, maker?: string) {
+        super(declarationToAbi('InvalidMakerError(bytes32 orderHash, address maker)'), {
+            orderHash,
+            maker,
+        });
+    }
+}
+
+export class FillError extends RichRevertReason {
+    constructor(orderHash?: string, error?: FillErrorCodes) {
+        super(declarationToAbi('FillError(bytes32 orderHash, uint8 error)'), { orderHash, error });
+    }
+}
+
+export class EpochOrderError extends RichRevertReason {
+    constructor(maker?: string, sender?: string, currentEpoch?: BigNumber | number | string) {
+        super(declarationToAbi('EpochOrderError(address maker, address sender, uint256 currentEpoch)'), {
+            maker,
+            sender,
+            currentEpoch,
+        });
+    }
+}
+
+export class AssetProxyExistsError extends RichRevertReason {
+    constructor(proxy?: string) {
+        super(declarationToAbi('AssetProxyExistsError(address proxy)'), { proxy });
+    }
+}
+
+export class AssetProxyDispatchError extends RichRevertReason {
+    constructor(error?: AssetProxyDispatchErrorCodes) {
+        super(declarationToAbi('AssetProxyDispatchError(uint8 error)'), { error });
+    }
+}
+
+export class NegativeSpreadError extends RichRevertReason {
+    constructor(leftOrderHash?: string, rightOrderHash?: string) {
+        super(declarationToAbi('NegativeSpreadError(bytes32 leftOrderHash, bytes32 rightOrderHash)'), {
+            leftOrderHash,
+            rightOrderHash,
+        });
+    }
+}
+
+export class TransactionExecutionError extends RichRevertReason {
+    constructor(error?: TransactionExecutionErrorCodes) {
+        super(declarationToAbi('TransactionExecutionError(uint8 error)'), { error });
+    }
+}
+
+export class IncompleteFillError extends RichRevertReason {
+    constructor(orderHash?: string) {
+        super(declarationToAbi('IncompleteFillError(bytes32 orderHash)'), { orderHash });
+    }
+}
+
+const RICH_REVERT_REGISTRY: RichRevertReason[] = [
+    new StandardError(),
+    new OrderStatusError(),
+    new SignatureError(),
+    new InvalidSenderError(),
+    new InvalidTakerError(),
+    new InvalidMakerError(),
+    new FillError(),
+    new EpochOrderError(),
+    new AssetProxyExistsError(),
+    new AssetProxyDispatchError(),
+    new NegativeSpreadError(),
+    new TransactionExecutionError(),
+    new IncompleteFillError(),
+];
+
+const RICH_REVERT_LUT: ObjectMap<RichRevertReason> = _.zipObject(
+    _.map(RICH_REVERT_REGISTRY, r => r.selector),
+    RICH_REVERT_REGISTRY,
+);
+
+const DECODER_CACHE: ObjectMap<Decoder> = _.zipObject(
```

Comment:
Our naming convention for Objects is stuffToStuff1ToStuff2 etc...
For example `selectorToDecoder`